### ISSUE TITLE
Include RapidJSON headers only if BOND_SIMPLE_JSON_PROTOCOL defined

### ISSUE THREAD

**Issue by oliora** [Microsoft/bond#44]
Correct me if I'm wrong but there is no reason for Bond user to include RapidJSON headers in case it doesn't need a JSON protocol. But the headers are included in any case so user got a dependency on RapidJSON even it's not needed.

**Comment by chwarr** [comment]
Yes, this makes sense.

**Comment by sapek** [comment]
The `boost::variant` [used](https://github.com/Microsoft/bond/blob/ac76e546ae0a6753a8f8defb76280e3610eff4f3/cpp/inc/bond/core/protocol.h#L202-L207) in the implementation of `bonded<T>` always has all protocols, even disabled ones. The reason for this was to minimize potential for catastrophic, potentially exploitable and very hard to diagnose errors when ODR rule violation would lead to wrong layout of the internal variant data.

**Comment by oliora** [comment]
I've checked the code. The `bonded` type that minded as protocol agnostic type in fact on all protocols. That's really strange for me. How this approach will allow to add new protocols without breaking binary compatibility with old code?

**Comment by chadwalters** [comment]
Revisiting this old issue. Many of the ODR-related concerns and the issue raised above about bonded<> are now addressed as of https://github.com/Microsoft/bond/pull/425. However, SimpleJSON still appears in the list of BuiltInProtocols. I think that this would be a fairly impactful breaking change for some of the heavier users of Bond internally.

**Comment by chwarr** [comment]
Agreed about the breakage. RapidJSON could be made opt-out pretty easily now, though.
